posts/react-compound-components/2021-5-25-React-compound-compnents.md

```diff
@@ -0,0 +1,250 @@
+
+```tsx
+const Accordion: React.FC<{
+  children: ReactNode | ReactNode[];
+  className?: string
+}> = ({ children, className }) => {
+  const [activeItem, setActiveItem] = useState("");
+  // function to update the active item
+   const changeActiveItem = useCallback(
+    (value) => {
+      if (activeItem !== value) setActiveItem(value);
+    },
+    [setActiveItem, activeItem]
+  );
+
+  return <div className={className}>{children}</div>
+}
+``` 
+
+we had created the accordion component, now we need to pass or share the state and the function to update the state to its children. To achieve this, we are going to use the **React Context**. If you are not familiar with React context, please refer  [https://reactjs.org/docs/context.html](https://reactjs.org/docs/context.html). Iam not gonna explain about react context here, but i will give you a hint about react context.
+
+> Context provides a way to pass data through the component tree without having to pass props down manually at every level.
+
+Now let's add context to our accordion component.
+
+```tsx
+import { createContext, useContext } from "react";
+
+// Creating the context for the Accordion.
+export const AccordionContext = createContext<{
+  activeItem: string;
+  changeSelectedItem: (item: string) => void;
+}>({ activeItem: "", changeSelectedItem: () => {} });
+
+export const useAccordionContext = () => {
+  const context = useContext(AccordionContext);
+  if (!context) {
+    throw new Error("Error in creating the context");
+  }
+  return context;
+};
+``` 
+
+After creating the context, we need to provide values to the context, it is done by **Context.Provider** element. The values refers to the data that is to be shared among all the components in the accordion. The **Provider** component accepts a value prop to be passed to consuming components that are descendants of this Provider. One Provider can be connected to many consumers.
+
+```tsx
+const Accordion: React.FC<{
+  children: ReactNode | ReactNode[];
+  className?: string;
+}> = ({ children, className }) => {
+  const [activeItem, setActiveItem] = useState("");
+
+  const changeActiveItem = useCallback(
+    (value) => {
+      if (activeItem !== value) setActiveItem(value);
+    },
+    [setActiveItem, activeItem]
+  );
+  return (
+    <AccordionContext.Provider
+      value={{ activeItem, changeSelectedItem: changeActiveItem }}
+    >
+      <div className={`accordion ${className}`}>{children}</div>
+    </AccordionContext.Provider>
+  );
+};
+
+export default Accordion;
+``` 
+
+In the accordion, we need to share the activeItem and changeSelectedItem to all the other components, so we are passing those two to the value of accordion context provider. Now We had built the root accordion component and also we have provided the values to the context. Let's build the remaining components and we are going to use the values consumed from the context and make the accordion component work as a whole.
```

Comment:
so we are passing them to the context provider which can be obtained using the the useAccordionContext custom hook (which uses the useContext hook internally).

Reply:
Let's build the remaining components. We are going to use the values consumed from the context and make the accordion component work as a whole.